Content.Client/_CD/Records/UI/RecordEntryEditPopup.xaml.cs

```diff
@@ -1,38 +1,145 @@
+        ValidationLabel.FontColorOverride = CharacterRecordViewer.ErrorColor;
+        ValidationLabel.Visible = false;
+
+        // Style the content panel
+        var styleBox = new StyleBoxFlat
+        {
+            BackgroundColor = CharacterRecordViewer.ContentPanelColor,
+            BorderColor = CharacterRecordViewer.BorderColor,
+            BorderThickness = new Thickness(1),
+        };
+
+        ContentPanel.PanelOverride = styleBox;
+
+        SetupEventHandlers();
+    }
+
+    private void SetupEventHandlers()
+    {
+        TitleEdit.OnTextChanged += args =>
+        {
+            if (args.Text.Length > PlayerProvidedCharacterRecords.TextMedLen)
+            {
+                TitleEdit.Text = args.Text[..PlayerProvidedCharacterRecords.TextMedLen];
+            }
+            ValidateFields();
+        };
+
+        InvolvedEdit.OnTextChanged += args =>
+        {
+            if (args.Text.Length > PlayerProvidedCharacterRecords.TextMedLen)
+            {
+                InvolvedEdit.Text = args.Text[..PlayerProvidedCharacterRecords.TextMedLen];
+            }
+            ValidateFields();
+        };
+
+        DescriptionEdit.OnTextChanged += _ => ValidateFields();
 
         SaveButton.OnPressed += _ =>
         {
+            if (!_isValid)
+                return;
             Close();
         };
+
+        DescriptionEdit.Placeholder =
+            new Rope.Leaf(Loc.GetString("cd-records-entry-edit-popup-description-placeholder"));
+
+        TitleEdit.PlaceHolder = Loc.GetString("cd-records-entry-edit-popup-title-placeholder");
+        InvolvedEdit.PlaceHolder = Loc.GetString("cd-records-entry-edit-popup-involved-placeholder");
+    }
+
+    private void ValidateFields()
+    {
+        var descriptionText = Rope.Collapse(DescriptionEdit.TextRope);
+        var descriptionLength = descriptionText.Length;
+
+        // Validation rules in priority order
+        // Overcomplicated into oblivion just because I didn't like how the else if statements looked
+        var rules = new[]
+        {
+            new ValidationRule(
+                () => string.IsNullOrWhiteSpace(TitleEdit.Text),
+                "cd-records-entry-edit-popup-title-required"),
+
+            new ValidationRule(
+                () => string.IsNullOrWhiteSpace(InvolvedEdit.Text),
+                "cd-records-entry-edit-popup-involved-required"),
+
+            new ValidationRule(
+                () => string.IsNullOrWhiteSpace(descriptionText),
+                "cd-records-entry-edit-popup-description-required"),
+
+            new ValidationRule(
+                () => descriptionLength > PlayerProvidedCharacterRecords.TextVeryLargeLen,
+                "cd-records-entry-edit-popup-description-too-long",
+                new[]
+                {
+                    ("current", (object)descriptionLength),
+                    ("max", PlayerProvidedCharacterRecords.TextVeryLargeLen)
+                }),
```

Comment:
It might be good to include a live-updating counter for that like paper has, however I would not consider it to be blocking a merge.

Reply:
i didn't include it because i couldn't find a good place for it, but this is going to show up the instant you go over the limit, so im not sure if there's much point to constantly displaying "x characters left"

Reply:
that is fine!